File: 김채린/추가문제/프로그래머스/택배 배달과 수거하기.java
import java.util.*;
class Solution {
    int clear;
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        long reuslt =0;
        int deliverPointer=n-1;
        int pickupsPointer = n-1;
        clear=deliveries.length;
        while(clear>0){
            if (deliveries[clear-1]==0&&pickups[clear-1]==0){
                clear--;
                continue;
            }
            reuslt+=clear*2;
            int deliver = cap;
            int pickup =cap;
            while (deliverPointer>=0&&deliver>=0){
                int a = deliveries[deliverPointer];
                if(deliveries[deliverPointer]<deliver){
                    deliver-=deliveries[deliverPointer];
                    deliveries[deliverPointer]=0;
                    deliverPointer--;
                } else if (a==deliver) {
                    deliveries[deliverPointer]=0;
                    break;
                } else{
                    deliveries[deliverPointer]-=deliver;
                    break;
                }
            }
            while (pickupsPointer>=0&&pickup>=0){
                int a =pickups[pickupsPointer];
                if(a<pickup){
                    pickup-=pickups[pickupsPointer];
                    pickups[pickupsPointer]=0;
                    pickupsPointer--;
                } else if (a==pickup) {
                    pickups[pickupsPointer]=0;
                    break;
                } else{
                    pickups[pickupsPointer]-=pickup;
                    break;
                }
            }
        }
        return reuslt;
    }
}

File: 이진영/프로그래머스/택배 배달과 수거하기.js
// 검색 풀이
function solution(cap, n, deliveries, pickups) {
    let answer = 0;
    
    let deliveries_amount = 0;
    let pickups_amount = 0;
    
    for (let i = n - 1; i > -1; i--) {
        deliveries_amount += deliveries[i];
        pickups_amount += pickups[i];
        
        let count = 0;
        while (deliveries_amount > 0 || pickups_amount > 0) {
            deliveries_amount -= cap;
            pickups_amount -= cap; // 음수가 될 경우 다음으로 넘김
            // 하나라도 양수이면 갔다 와야한다.
            count++;
        }
        answer += (i + 1) * 2 * count;
    }
    
    return answer;
}